docs/proposals/20210222-azure-machinepool-machine.md

```diff
@@ -0,0 +1,332 @@
+AzureMachinePoolInstanceStatus struct {
+    // Version defines the Kubernetes version for the VM Instance
+    // +optional
+    Version string `json:"version"`
+    
+    // ProvisioningState is the provisioning state of the Azure virtual machine instance.
+    // +optional
+    ProvisioningState *infrav1.VMState `json:"provisioningState"`
+    
+    // ProviderID is the provider identification of the VMSS Instance
+    // +optional
+    ProviderID string `json:"providerID"`
+    
+    // InstanceID is the identification of the Machine Instance within the VMSS
+    // +optional
+    InstanceID string `json:"instanceID"`
+    
+    // InstanceName is the name of the Machine Instance within the VMSS
+    // +optional
+    InstanceName string `json:"instanceName"`
+    
+    // LatestModelApplied indicates the instance is running the most up-to-date VMSS model. A VMSS model describes
+    // the image version the VM is running. If the instance is not running the latest model, it means the instance
+    // may not be running the version of Kubernetes the Machine Pool has specified and needs to be updated.
+    LatestModelApplied bool `json:"latestModelApplied"`
+}
+```
+
+#### Proposed API Changes
+The proposed changes below show the AzureMachinePool and AzureMachinePoolMachine.
+
+```go
+// AzureMachinePoolSpec defines the desired state of AzureMachinePool
+AzureMachinePoolSpec struct {
+
+    /*
+       Other fields omitted for brevity
+    */
+	
+    // MaxUnavailable is the max number of replicas which can be unavailable at any time. For example, if the
+    // machine pool has a replica count of 5 and a MaxUnavailable value of 2, a rolling upgrade will never exceed
+    // more than 2 machines unavailable at a time.
+    //
+    // Note: If the MaxUnavailable value is equal to the machine pool replica value, there is
+    // the possibility of application down time.
+    // +kubebuilder:default=1
+    MaxUnavailable int32 `json:"maxUnavailable,omitempty"`
+
+    // MaxSurge is the number of replicas to surge during an upgrade. For example, if the machine pool has a replica
+    // count of 5 and a MaxSurge value of 4, during an upgrade, the number of Virtual Machine Scale Set instances
+    // will increase to 9 consisting of the 5 existing replicas and 4 new replicas running the updated model. The
+    // surge will temporaly increase the number of machines in the VMSS to expedite upgrades.
+    //
+    // Note: Azure Subscriptions have a limited quota of CPUs. This quota can be expanded upon request. Due to the
+    // nature of this feature, spikes due to surge may cause you to exceed quota.
+    //
+    // See Also: https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits
+    MaxSurge *int32 `json:"maxSurge,omitempty"`
+}
+
+// AzureMachinePoolMachineSpec defines the desired state of AzureMachinePoolMachine
+AzureMachinePoolMachineSpec struct {
```

Comment:
How do we make it clear to the users that they are not responsible for creating the AzureMachinePoolMachines?

Reply:
That's a great question. Do you have any suggestions?

My first inclination is to add documentation.

There really isn't anything that stops someone from creating an AMPM. It won't hurt anything.